pip_api/_parse_requirements.py

```diff
@@ -187,12 +397,51 @@ def parse_requirements(
             req: Optional[Union[requirements.Requirement, UnparsedRequirement]] = None
             known, _ = parser.parse_known_args(line.strip().split())
             if known.req:
-                # We want to stop combining strings before an @ sign with space on either side
-                req_str = str()
-                for r in known.req:
-                    if r == "@":
-                        break
-                    req_str += r
+                # TODO(alex): Extract this out to a separate helper?
+                # Strip out the marker temporarily while we parse out any potential URLs
+                req_str = str().join(known.req)
+                marker_str = None
+                link = None
+                if ";" in req_str:
+                    req_str, marker_str = req_str.split(";", 1)
+
+                if _is_url(req_str):
+                    link = Link(req_str)
+                else:
+                    path = os.path.normpath(os.path.abspath(req_str))
+                    p, _ = _strip_extras(path)
+                    url = _get_url_from_path(p, req_str)
+                    if url is not None:
+                        link = Link(url)
+
+                # it's a local file, dir, or url
+                if link is not None:
+                    # Handle relative file URLs
+                    if link.scheme == "file" and re.search(r"\.\./", link.url):
+                        link = Link(
+                            _path_to_url(os.path.normpath(os.path.abspath(link.path)))
+                        )
+                    # wheel file
+                    if link.is_wheel:
+                        wheel_info = WHEEL_FILE_RE.match(link.filename)
+                        wheel_name = wheel_info.group("name").replace("_", "-")
+                        wheel_version = wheel_info.group("ver").replace("_", "-")
+                        req_str = f"{wheel_name}=={wheel_version}"
+                    else:
+                        # set the req to the egg fragment.  when it's not there, this
+                        # will become an 'unnamed' requirement
+                        #
+                        # TODO(alex): This should be using `_egg_fragment(link.url)`
+                        # But after we generate a URL via `_get_url_from_path`, the URL ends with
+                        # `egg%3Dfoo` rather than 'egg=foo` like the regex expects. This might be a
+                        # `pip` bug? Just keep things running with the original requirement string
+                        # while I figure this out.
+                        req_str = _egg_fragment(req_str)
+                        assert req_str is not None
+
+                # Reassemble the requirement string with the original marker
```

Comment:
The original `pip` code doesn't attach the marker to the `Requirement` and instead passes it back via a `RequirementsPart` structure. I think we want the marker in the returned `Requirement` so I'm tacking the marker back on once we're parsed the requirement string out of the URL.